sdk/dotnet/Docker.cs

```diff
@@ -0,0 +1,644 @@
+        /// in which the Dockerfile resides (although dockerfile may be used to choose a custom location
+        /// independent of this choice). If not specified, the context defaults to the current working
+        /// directory; if a relative path is used, it is relative to the current working directory that
+        /// Pulumi is evaluating.
+        /// </summary>
+        [Input("context")]
+        public Input<string>? Context { get; set; }
+
+        /// <summary>
+        /// <see cref="Dockerfile"/> may be used to override the default Dockerfile name and/or location.
+        /// By default, it is assumed to be a file named Dockerfile in the root of the build context.
+        /// </summary>
+        [Input("dockerfile")]
+        public Input<string>? Dockerfile { get; set; }
+
+        /// <summary>
+        /// An optional map of named build-time argument variables to set during the Docker build.
+        /// This flag allows you to pass built-time variables that can be accessed like environment variables
+        /// inside the `RUN` instruction.
+        /// </summary>
+        [Input("args")]
+        public InputMap<string>? Args { get; set; }
+
+        /// <summary>
+        /// An optional <see cref="CacheFrom"/> object with information about the build stages to use for the Docker
+        /// build cache.  This parameter maps to the --cache-from argument to the Docker CLI. If this
+        /// parameter is `true`, only the final image will be pulled and passed to --cache-from; if it is
+        /// a <see cref="CacheFrom"/> object, the stages named therein will also be pulled and passed to --cache-from.
+        /// </summary>
+        [Input("cacheFrom")]
+        public InputUnion<bool, CacheFrom>? CacheFrom { get; set; }
+
+        /// <summary>
+        /// An optional catch-all string to provide extra CLI options to the docker build command.
+        /// For example, use to specify `--network host`.
+        /// </summary>
+        [Input("extraOptions")]
+        public Input<Input<string>[]>? ExtraOptions { get; set; }
+
+        /// <summary>
+        /// Environment variables to set on the invocation of `docker build`, for example to support
+        /// `DOCKER_BUILDKIT=1 docker build`.
+        /// </summary>
+        [Input("env")]
+        public Dictionary<string, string>? Env { get; set; }
+    }
+
+    /// <summary>
+    /// A copy of <see cref="DockerBuild"/> but without using <see cref="Input{T}"/> in types.
+    /// </summary>
+    internal class DockerBuildUnwrap
+    {
+        public string? Context { get; set; }
+        public string? Dockerfile { get; set; }
+        public ImmutableDictionary<string, string>? Args { get; set; }
+        public CacheFromUnwrap? CacheFrom { get; set; }
+        public ImmutableArray<string>? ExtraOptions { get; set; }
+        public Dictionary<string, string>? Env { get; set; }
```

Comment:
can we consistently use immutable data?

Reply:
Changed everywhere in arguments and unwraps